experimental/type-checker/src/KindChecker.hs

```diff
@@ -0,0 +1,339 @@
+{-# OPTIONS_GHC -Wno-missing-signatures #-}
+{-# OPTIONS_GHC -Wno-unused-top-binds #-}
+
+module KindChecker (getType) where
+
+import Control.Monad.Trans.Reader (ReaderT, runReaderT)
+import Control.Monad.Trans.State (StateT, evalStateT)
+import Control.Monad.Trans.Writer (WriterT (runWriterT))
+import Data.Bifunctor (Bifunctor (second))
+
+import Control.Monad.Except
+import Control.Monad.Reader.Class
+import Control.Monad.State.Class
+import Control.Monad.Writer.Class
+
+import Prettyprinter
+
+type Atom = String
+infixr 8 :->:
+
+data Kind
+  = Kind
+  | Kind :->: Kind
+  | KVar String
+  deriving (Eq, Show)
+
+instance Pretty Kind where
+  pretty = \case
+    Kind -> pretty "*"
+    ((x :->: y) :->: z) -> parens (pretty $ x :->: y) <+> pretty "→" <+> pretty z
+    x :->: y -> pretty x <+> pretty "→" <+> pretty y
+    KVar a -> pretty a
+
+data Type
+  = Var Atom
+  | App Type Type
```

Comment:
I think it's fine like it is? Using a list seems like it would give us two different ways of representing the same thing, i.e. 

`App IntT (App StringT EitherT)` could be `App [IntT,StringT] EitherT` or `App [IntT] (App [StringT] EitherT)` (I flipped the order b/c it seems more natural to me for the args to be on the left here)

And then we have to worry about catching "empty applications", which is just one more thing that can go wrong.